simulator/src/scrypto_bindgen/translation.rs

```diff
@@ -0,0 +1,310 @@
+                            blueprint_name,
+                            schema_resolver
+                        )?
+                    ))
+                }
+                (Some(name), ..) => Ok(name.to_owned()),
+                (None, ..) => Err(schema::SchemaError::NoNameFound),
+            }?
+        }
+        TypeKind::Map {
+            key_type,
+            value_type,
+        } => type_ident.unwrap_or(format!(
+            "IndexMap<{}, {}>",
+            type_name(
+                &ScopedTypeId(type_identifier.0, key_type),
+                blueprint_name,
+                schema_resolver
+            )?,
+            type_name(
+                &ScopedTypeId(type_identifier.0, value_type),
+                blueprint_name,
+                schema_resolver
+            )?
+        )),
+        TypeKind::Custom(ScryptoCustomTypeKind::Decimal) => "Decimal".to_owned(),
+        TypeKind::Custom(ScryptoCustomTypeKind::PreciseDecimal) => "PreciseDecimal".to_owned(),
+        TypeKind::Custom(ScryptoCustomTypeKind::NonFungibleLocalId) => {
+            "NonFungibleLocalId".to_owned()
+        }
+
+        TypeKind::Custom(ScryptoCustomTypeKind::Own)
+            if type_validation
+                == TypeValidation::Custom(ScryptoCustomTypeValidation::Own(
+                    OwnValidation::IsBucket,
+                )) =>
+        {
+            "Bucket".to_owned()
+        }
+        TypeKind::Custom(ScryptoCustomTypeKind::Own)
+            if type_validation
+                == TypeValidation::Custom(ScryptoCustomTypeValidation::Own(
+                    OwnValidation::IsProof,
+                )) =>
+        {
+            "Proof".to_owned()
+        }
+        TypeKind::Custom(ScryptoCustomTypeKind::Own)
+            if type_validation
+                == TypeValidation::Custom(ScryptoCustomTypeValidation::Own(
+                    OwnValidation::IsVault,
+                )) =>
+        {
+            "Vault".to_owned()
+        }
+        TypeKind::Custom(ScryptoCustomTypeKind::Own)
+            if type_validation
+                == TypeValidation::Custom(ScryptoCustomTypeValidation::Own(
+                    OwnValidation::IsKeyValueStore,
+                )) =>
+        {
+            "NodeId".to_owned()
+        }
+        TypeKind::Custom(ScryptoCustomTypeKind::Own)
+            if type_validation
+                == TypeValidation::Custom(ScryptoCustomTypeValidation::Own(
+                    OwnValidation::IsGlobalAddressReservation,
+                )) =>
+        {
+            "GlobalAddressReservation".to_owned()
+        }
+        TypeKind::Custom(ScryptoCustomTypeKind::Own) => match type_validation {
+            TypeValidation::Custom(ScryptoCustomTypeValidation::Own(
+                OwnValidation::IsTypedObject(package_address, bp_name),
+            )) if package_address.is_none()
+                || package_address == Some(schema_resolver.package_address()) =>
```

Comment:
We should look up the address from database?

Also, minor, the `None` is unexpected (again, hopefully covered by tests).

Reply:
Goot to know that the `None` case is not expected. I thought that a package address of `None` meant "this package".

About looking up the address, why would we want to look it up from the database? Is the check that the passed `PackageAddress` is equal to the one in the `TypeValidation` not sufficient?

Reply:
I was wrong, despite implementing it myself. :(

`None` means the source package address **IF** the schema originates from a blueprint.

Can you revert the last change about this?

Reply:
No worries! Will revert now